File: BW/11월1주차/땅따먹기.js
function solution(land) {
    var answer = 0;
    
    for (var i = 1; i < land.length; i++) {
        // 현재 행 + 이전 행 중 같은 인덱스를 제외한 최댓값
        land[i][0] += Math.max(land[i - 1][1], land[i - 1][2], land[i - 1][3]);
        land[i][1] += Math.max(land[i - 1][0], land[i - 1][2], land[i - 1][3]);
        land[i][2] += Math.max(land[i - 1][0], land[i - 1][1], land[i - 1][3]);
        land[i][3] += Math.max(land[i - 1][0], land[i - 1][1], land[i - 1][2]);
    }

    // 마지막 행 최대값
    answer = Math.max(...land[land.length - 1]);
    
    return answer;
}

File: JY/11월 1주차/땅따먹기.py
import sys
sys.setrecursionlimit(10**8)
MAX_ROW = 4

def find_max_value(land, column, target):
    max_value = 0
    for row in range(MAX_ROW):
        if row == target:
            continue
        max_value = max(max_value, land[column][row])
    return max_value

def solution(land):
    answer = 0
    dp = [[0] *MAX_ROW for _ in range(len(land))]

    # 테이블 초기화
    for row in range(MAX_ROW):
        dp[0][row] = land[0][row]
    
    # 1번 행부터 갱신 
    for column in range(1, len(land)): # N
        # column-1(이전 행)의 최댓값 and row(열)이 다른 값 + 자기 자신
        for row in range(MAX_ROW): # 4
            dp[column][row] = land[column][row] + find_max_value(dp, column-1, row) #4
            
    answer = max(dp[len(land)-1])
    return answer